Spec for MSC3077 (#1602)
matrix-org/matrix-spec: content/client-server-api/modules/voip_events.md

> ##### Streams

Reviewer: can we add something to https://pr1602--matrix-spec-previews.netlify.app/client-server-api/#common-fields about the new field?

Clients may send more than one stream in a VoIP call. Metadata can be included in
the `m.call.invite`, `m.call.answer` and `m.call.negotiate` events to describe the

> streams being sent. This is the `sdp_stream_metadata` field.

Reviewer: please prefer 'property' over 'field': https://github.com/matrix-org/matrix-spec/blob/main/meta/documentation_style.rst#general

> streams being sent. This is the `sdp_stream_metadata` field.

Reviewer: ```suggestion
streams being sent, using the `sdp_stream_metadata` property.
```

This field is an object in which each key is one stream `id` in the session
description. The values are objects with the following fields:

 * `purpose` - a string indicating the purpose of the stream. For compatibility

>    between client the following values are defined:

Reviewer: why "For compatibility between client [sic]" ?

> This field is an object in which each key is one stream `id` in the session
> description. The values are objects with the following fields:
> 
>  * `purpose` - a string indicating the purpose of the stream. For compatibility
>    between client the following values are defined:

Reviewer: this is pretty hard to grok. It doesn't need a formal description of the structure here; that is provided by the event definition tables. Better to be informal and easier to parse.

```suggestion
`sdp_stream_metadata` maps from the `id` of a stream in the session description, 
to metadata about that stream. 

Currently only one property is defined for the metadata. This is `purpose`, which should be a stringindicating the purpose of the stream. The following `purpose`s are defined:
```

   *  `m.usermedia` - stream that contains the webcam and/or microphone tracks
   *  `m.screenshare` - stream with the screen-sharing tracks

If an incoming stream is not described in `sdp_stream_metadata` and
`sdp_stream_metadata` is present, the stream should be ignored. If a stream has
a `purpose` of an unknown type (i.e. not `m.usermedia` or `m.screenshare`), it
should be ignored.

Clients implementing this specification will ignore any streamless tracks. Note
that in the JavaScript WebRTC API, this means `addTrack()` must be passed two
parameters: a track and a stream, not just a track, and in a video call the

> stream must be the same for both audio and video track.

Reviewer: I know this is existing text, but it now seems even more confusing than before, so I think we need to clarify.

1. What is the mechanism by which streamless tracks are ignored? Or is this an *instruction* to clients to ignore  such tracks?
2. what is that to do with `addTrack`'s parameters?
3. is it related to `sdp_stream_metadata`? if not, it should probably not be in the middle of the text about it?

> During the initial invite and answer exchange clients find out if the field
> `sdp_stream_metadata` is missing. For backwards compatibility if it is not
> present in the event sent by the opponent, the client should ignore any new

Reviewer: ```suggestion
For backwards compatibility, if `sdp_stream_metadata` is not present in the
initial `m.call.invite` or `m.call.answer` [please link] event sent by the other party, 
the client should ignore any new
```

(I'm not entirely following why it is necessary to ignore new streams, but that is certainly what the MSC says)

incoming streams (i.e. it should use the first one) and it shouldn't send more
than one stream (i.e. clients cannot send a video feed and a screenshare at the

> same time, as is the case in current clients).

Reviewer: inappropriate outside an MSC

```suggestion
same time).
```
